src/dali/plugin/input/rest/kucoin_futures.py

```diff
@@ -0,0 +1,288 @@
+        return None
+
+    def _process_gains(
+        self,
+        in_transactions: List[InTransaction],
+        out_transactions: List[OutTransaction],
+    ) -> None:
+        pass
+
+    def _fetch_data(
+        self,
+        start_at: int,
+        end_at: int,
+    ) -> List[Dict[str, Any]]:
+
+        offset: int = 1
+        retries: int = 0
+        has_more: bool = False
+
+        ledger_transactions: Dict[str, Any] = {}
+        items: List[Dict[str, Any]] = []
+
+        while (start_at <= end_at) and retries < 4:
+            try:
+                ledger_transactions = self._client.futuresPrivateGetTransactionHistory({_START_AT: start_at, _END_AT: start_at + _TS_INCREMENT})
+                data_list = ledger_transactions[_DATA][_DATA_LIST]
+                has_more = ledger_transactions[_DATA][_HAS_MORE]
+
+                for item in data_list:
+                    items.append(item)
+                    offset = item[_OFFSET]
+
+                # https://github.com/ccxt/ccxt/issues/10273
+                # enableRateLimit is not working for kucoin
+                time.sleep(0.15)
+                retries = 0
+
+                while has_more and (retries < 4):
+                    try:
+                        ledger_transactions = self._client.futuresPrivateGetTransactionHistory(
+                            {_START_AT: start_at, _END_AT: start_at + _TS_INCREMENT, _OFFSET: offset}
+                        )
+                        data_list = ledger_transactions[_DATA][_DATA_LIST]
+                        has_more = ledger_transactions[_DATA][_HAS_MORE]
+
+                        for item in data_list:
+                            items.append(item)
+                            offset = item[_OFFSET]
+
+                        time.sleep(0.15)
+
+                    except RateLimitExceeded:
+                        self._client.sleep(13000)
+                        retries += 1
+
+                if retries >= 4:
+                    raise Exception("Failed to fetch ledger transactions after 4 retries")
+
+                start_at += _TS_INCREMENT
+                offset = 1
+                has_more = False
+                retries = 0
+
+            except RateLimitExceeded:
+                self._client.sleep(13000)
+                retries += 1
+
+            except Exception as exception:
+                raise exception
+
+        if retries >= 4:
+            raise Exception("Failed to fetch ledger transactions after 4 retries")
+
+        return items
+
+    def _process_data(
+        self,
+        items: List[Dict[str, Any]],
+        in_transactions: List[InTransaction],
+        out_transactions: List[OutTransaction],
+        # intra_transactions: List[IntraTransaction],
+    ) -> None:
+        for item in items:
+            timestamp_value: int = item[_TIME]
+            timestamp: str = datetime.fromtimestamp(float(timestamp_value) / 1000.0, tz.tzutc()).strftime("%Y-%m-%d %H:%M:%S.%f%z")
```

Comment:
Use the `self._rp2_timestamp_from_ms_epoch()` function to convert millisecond epoch to RP2 timestamps:
```
self._rp2_timestamp_from_ms_epoch(item[_TIME])
```

Reply:
fixed